File: sorting/merge_sort.cc
#include "utils/test.h"

namespace algorithm {
namespace sorting {

template <typename T, std::size_t N>
void Merge(T (&array)[N], int32_t lo, int32_t mi, int32_t hi) {
  auto l_len = mi - lo + 1, r_len = hi - mi;
  auto *l_arr = new T[l_len], *r_arr = new T[r_len];

  for (auto i = 0; i < l_len; i++) {
    l_arr[i] = array[lo + i];
  }
  for (auto i = 0; i < r_len; i++) {
    r_arr[i] = array[mi + 1 + i];
  }

  auto i = 0, j = 0, k = lo;
  while (i < l_len && j < r_len) {
    array[k++] = l_arr[i] <= r_arr[j] ? l_arr[i++] : r_arr[j++];
  }
  while (i < l_len) {
    array[k++] = l_arr[i++];
  }
  while (j < r_len) {
    array[k++] = r_arr[j++];
  }

  delete[] l_arr;
  delete[] r_arr;
}

template <typename T, std::size_t N>
void MergeSort(T (&array)[N], int32_t lo, int32_t hi) {
  if (lo < hi) {
    int32_t mi = (lo + hi) / 2;
    MergeSort(array, lo, mi);
    MergeSort(array, mi + 1, hi);
    Merge(array, lo, mi, hi);
  }
}

namespace test {

void MergeSort_TestHandlesMultipleTypesOfArrayInput() {
  int32_t ints[11] = {6, 8, 10, 26, 9, 2, 40, 22, 5, 32, 3};
  std::string strs[7] = {"c", "java", "cpp", "python", "go", "sql", "swift"};
  float flts[8] = {5.2f, 4.4f, 6.8f, 9.5f, 0.7f, 1.2f, 3.2f, 5.9f};

  int32_t sorted_ints[11] = {2, 3, 5, 6, 8, 9, 10, 22, 26, 32, 40};
  std::string sorted_strs[7] = {"c",      "cpp", "go",   "java",
                                "python", "sql", "swift"};
  float sorted_flts[8] = {0.7f, 1.2f, 3.2f, 4.4f, 5.2f, 5.9f, 6.8f, 9.5f};

  MergeSort(ints, 0, std::size(ints) - 1);
  EXPECT_ARR_EQ(ints, sorted_ints);

  MergeSort(strs, 0, std::size(strs) - 1);
  EXPECT_ARR_EQ(strs, sorted_strs);

  MergeSort(flts, 0, std::size(flts) - 1);
  EXPECT_ARR_EQ(flts, sorted_flts);
}

void RunTests() { TEST(MergeSort_TestHandlesMultipleTypesOfArrayInput); }

}  // namespace test

}  // namespace sorting
}  // namespace algorithm

int main() {
  std::cout << "-----Running merge sort tests-----" << std::endl;
  algorithm::sorting::test::RunTests();
  return 0;
}

File: sorting/merge_sort.py
"""
Merge Sort
"""
import unittest
from typing import TypeVar

T = TypeVar("T")


def merge_sort(array: list[T], lo: int, hi: int):
    if lo < hi:
        mi: int = (lo + hi) // 2
        merge_sort(array, lo, mi)
        merge_sort(array, mi + 1, hi)
        merge(array, lo, mi, hi)


def merge(array: list[T], lo: int, mi: int, hi: int):
    l_len, r_len = mi - lo + 1, hi - mi
    l_arr, r_arr = [None] * l_len, [None] * r_len

    for i in range(l_len):
        l_arr[i] = array[lo + i]
    for i in range(r_len):
        r_arr[i] = array[mi + 1 + i]

    i, j, k = 0, 0, lo
    while i < l_len and j < r_len:
        if l_arr[i] > r_arr[j]:
            array[k] = r_arr[j]
            j += 1
        else:
            array[k] = l_arr[i]
            i += 1
        k += 1

    while i < l_len:
        array[k] = l_arr[i]
        k += 1
        i += 1

    while j < r_len:
        array[k] = r_arr[j]
        k += 1
        j += 1


class TestMergeSort(unittest.TestCase):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.integers = [6, 8, 10, 26, 9, 2, 40, 22, 5, 32, 3]
        self.strings = ["c", "assembly", "cpp", "python", "go", "swift", "sql"]
        self.floats = [5.2, 4.4, 6.8, 9.5, 0.7, 1.2, 3.2, 5.9]

    def test_handles_multiple_array_type_input(self):
        merge_sort(self.integers, 0, len(self.integers) - 1)
        self.assertListEqual(self.integers, [2, 3, 5, 6, 8, 9, 10, 22, 26, 32, 40])
        merge_sort(self.strings, 0, len(self.strings) - 1)
        self.assertListEqual(
            self.strings, ["assembly", "c", "cpp", "go", "python", "sql", "swift"]
        )
        merge_sort(self.floats, 0, len(self.floats) - 1)
        self.assertListEqual(self.floats, [0.7, 1.2, 3.2, 4.4, 5.2, 5.9, 6.8, 9.5])


if __name__ == "__main__":
    unittest.main()
